physlr/physlr.py

```diff
@@ -400,14 +400,49 @@ def diameter_of_tree(g, weight=None):
         return (u, v, diameter)
 
     @staticmethod
-    def determine_backbones_of_trees(g):
-        "Determine the backbones of the maximum spanning trees."
+    def detect_junctions_of_tree(gcomponent, messages, junction_threshold):
+        """"
+        detect the junctions in the tree,
+        with more than 2 branches larger than junction_threshold.
+        """
+        candidate_junctions = {node
+                               for node in list(gcomponent.nodes)
+                               if gcomponent.degree(node) > 2}
+        junctions = []
+        for candidate in candidate_junctions:
+            candidate_messages = [messages[(candidate, neighbor)]
+                                  for neighbor in gcomponent.neighbors(candidate)]
+            candidate_messages.sort()
+            if candidate_messages[-3] > junction_threshold:
+                junctions.append(candidate)
+        return junctions
+
+    @staticmethod
+    def determine_backbones_of_trees(g, junction_threshold=0):
+        """"
+        Determine the backbones of the maximum spanning trees
+        and remove junctions of more than 2 branches larger than junction_threshold.
+        """
         paths = []
-        for component in nx.connected_components(g):
-            gcomponent = g.subgraph(component)
-            u, v, _ = Physlr.diameter_of_tree(gcomponent, weight="n")
-            path = nx.shortest_path(gcomponent, u, v, weight="n")
-            paths.append(path)
+        if junction_threshold > 0:
+            for component in nx.connected_components(g):
+                gcomponent = g.subgraph(component)
+                messages = Physlr.determine_reachability_by_message_passing(gcomponent)
+                nodes_to_remove = \
+                    Physlr.detect_junctions_of_tree(gcomponent, messages, junction_threshold)
+                gcomponents = gcomponent.copy()
+                gcomponents.remove_nodes_from(nodes_to_remove)
+                for component2 in nx.connected_components(gcomponents):
+                    gcomponent2 = g.subgraph(component2)
+                    u, v, _ = Physlr.diameter_of_tree(gcomponent2, weight="n")
+                    path = nx.shortest_path(gcomponent2, u, v, weight="n")
+                    paths.append(path)
```

Comment:
These lines 435–439 are a perfect duplication of 441–445. Avoid copy-and-pasting code.